<center>
<h1><b>Big Data Analytics Project</b></h1>
<h3><b>Credit Card Fraud Detection</b></h3>
</center>
By:
SOUBHIK SINHA (19BIT0303)
AASHISH BANSAL (19BIT0346)

Component: DECISION TREE

# Importing Libraries

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
import pandas as ps

from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import confusion_matrix
from sklearn.metrics import classification_report

# Importing Dataset

In [ ]:
cred = ps.read_csv("/content/drive/MyDrive/Project - ITE2013 - Big Data - Credit Card Fraud Detection/dataset/creditcard.csv")

In [ ]:
cred

Time         V1         V2  ...       V28  Amount  Class
0            0.0  -1.359807  -0.072781  ... -0.021053  149.62      0
1            0.0   1.191857   0.266151  ...  0.014724    2.69      0
2            1.0  -1.358354  -1.340163  ... -0.059752  378.66      0
3            1.0  -0.966272  -0.185226  ...  0.061458  123.50      0
4            2.0  -1.158233   0.877737  ...  0.215153   69.99      0
...          ...        ...        ...  ...       ...     ...    ...
284802  172786.0 -11.881118  10.071785  ...  0.823731    0.77      0
284803  172787.0  -0.732789  -0.055080  ... -0.053527   24.79      0
284804  172788.0   1.919565  -0.301254  ... -0.026561   67.88      0
284805  172788.0  -0.240440   0.530483  ...  0.104533   10.00      0
284806  172792.0  -0.533413  -0.189733  ...  0.013649  217.00      0

[284807 rows x 31 columns]

# Shape of Dataset

In [ ]:
print(f"Dataset Shape : \n {cred.shape}")

Dataset Shape : 
 (284807, 31)


# Exploratory Data Analysis

In [ ]:
print(cred.describe())

                Time            V1  ...         Amount          Class
count  284807.000000  2.848070e+05  ...  284807.000000  284807.000000
mean    94813.859575  1.758743e-12  ...      88.349619       0.001727
std     47488.145955  1.958696e+00  ...     250.120109       0.041527
min         0.000000 -5.640751e+01  ...       0.000000       0.000000
25%     54201.500000 -9.203734e-01  ...       5.600000       0.000000
50%     84692.000000  1.810880e-02  ...      22.000000       0.000000
75%    139320.500000  1.315642e+00  ...      77.165000       0.000000
max    172792.000000  2.454930e+00  ...   25691.160000       1.000000

[8 rows x 31 columns]


In [ ]:
cred.head()

Time        V1        V2        V3  ...       V27       V28  Amount  Class
0   0.0 -1.359807 -0.072781  2.536347  ...  0.133558 -0.021053  149.62      0
1   0.0  1.191857  0.266151  0.166480  ... -0.008983  0.014724    2.69      0
2   1.0 -1.358354 -1.340163  1.773209  ... -0.055353 -0.059752  378.66      0
3   1.0 -0.966272 -0.185226  1.792993  ...  0.062723  0.061458  123.50      0
4   2.0 -1.158233  0.877737  1.548718  ...  0.219422  0.215153   69.99      0

[5 rows x 31 columns]

In [ ]:
print(f"Columns / Feature / Variable names : \n {cred.columns}")

Columns / Feature / Variable names : 
 Index(['Time', 'V1', 'V2', 'V3', 'V4', 'V5', 'V6', 'V7', 'V8', 'V9', 'V10',
       'V11', 'V12', 'V13', 'V14', 'V15', 'V16', 'V17', 'V18', 'V19', 'V20',
       'V21', 'V22', 'V23', 'V24', 'V25', 'V26', 'V27', 'V28', 'Amount',
       'Class'],
      dtype='object')


In [ ]:
print(f"Unique values of target variable / Attribute : \n {cred['Class'].unique()}")

Unique values of target variable / Attribute : 
 [0 1]


In [ ]:
# 0 means valid , 1 means fraudulent

In [ ]:
print(f"Number of samples under each target value (0 and 1) : \n {cred['Class'].value_counts()}")

Number of samples under each target value (0 and 1) : 
 0    284315
1       492
Name: Class, dtype: int64


# Feature Engineering

## Removing unneeded Features

In [ ]:
cred = cred.drop(['Time'], axis=1)
print(f"Features left after removal of 'Time' feature : \n{cred.columns}")

Features left after removal of 'Time' feature : 
Index(['V1', 'V2', 'V3', 'V4', 'V5', 'V6', 'V7', 'V8', 'V9', 'V10', 'V11',
       'V12', 'V13', 'V14', 'V15', 'V16', 'V17', 'V18', 'V19', 'V20', 'V21',
       'V22', 'V23', 'V24', 'V25', 'V26', 'V27', 'V28', 'Amount', 'Class'],
      dtype='object')


## Checking for NULL or NaN values

In [ ]:
print(f"Dataset information : \n {cred.info()}")

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 284807 entries, 0 to 284806
Data columns (total 30 columns):
 #   Column  Non-Null Count   Dtype  
---  ------  --------------   -----  
 0   V1      284807 non-null  float64
 1   V2      284807 non-null  float64
 2   V3      284807 non-null  float64
 3   V4      284807 non-null  float64
 4   V5      284807 non-null  float64
 5   V6      284807 non-null  float64
 6   V7      284807 non-null  float64
 7   V8      284807 non-null  float64
 8   V9      284807 non-null  float64
 9   V10     284807 non-null  float64
 10  V11     284807 non-null  float64
 11  V12     284807 non-null  float64
 12  V13     284807 non-null  float64
 13  V14     284807 non-null  float64
 14  V15     284807 non-null  float64
 15  V16     284807 non-null  float64
 16  V17     284807 non-null  float64
 17  V18     284807 non-null  float64
 18  V19     284807 non-null  float64
 19  V20     284807 non-null  float64
 20  V21     284807 non-null  float64
 21  V22     28

# Data Transformation

In [ ]:
print(f"Some Amount Column / Feature values : \n {cred['Amount'][0:4]}")

Some Amount Column / Feature values : 
 0    149.62
1      2.69
2    378.66
3    123.50
Name: Amount, dtype: float64


# Data Preprocessing

In [ ]:
cred['norm_amount'] = StandardScaler().fit_transform(cred['Amount'].values.reshape(-1,1))
cred = cred.drop(['Amount'], axis=1)
print(f"Some Amount Couln values - After application of StandardScalar : \n {cred['norm_amount'][0:4]}")

Some Amount Couln values - After application of StandardScalar : 
 0    0.244964
1   -0.342475
2    1.160686
3    0.140534
Name: norm_amount, dtype: float64


# Creation of features and target
Here, the "Class" column is dropped because it is the column which contains the result and we need to predict it, so we cannot use it as a feature.

In [ ]:
X = cred.drop(['Class'], axis=1)
Y = cred[['Class']]

In [ ]:
X

V1         V2        V3  ...       V27       V28  norm_amount
0       -1.359807  -0.072781  2.536347  ...  0.133558 -0.021053     0.244964
1        1.191857   0.266151  0.166480  ... -0.008983  0.014724    -0.342475
2       -1.358354  -1.340163  1.773209  ... -0.055353 -0.059752     1.160686
3       -0.966272  -0.185226  1.792993  ...  0.062723  0.061458     0.140534
4       -1.158233   0.877737  1.548718  ...  0.219422  0.215153    -0.073403
...           ...        ...       ...  ...       ...       ...          ...
284802 -11.881118  10.071785 -9.834783  ...  0.943651  0.823731    -0.350151
284803  -0.732789  -0.055080  2.035030  ...  0.068472 -0.053527    -0.254117
284804   1.919565  -0.301254 -3.249640  ...  0.004455 -0.026561    -0.081839
284805  -0.240440   0.530483  0.702510  ...  0.108821  0.104533    -0.313249
284806  -0.533413  -0.189733  0.703337  ... -0.002415  0.013649     0.514355

[284807 rows x 29 columns]

In [ ]:
Y

Class
0           0
1           0
2           0
3           0
4           0
...       ...
284802      0
284803      0
284804      0
284805      0
284806      0

[284807 rows x 1 columns]

# Splitting the dataset for training and testing

In [ ]:
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size=0.3, random_state=0)
print(X_train.shape)
print(X_test.shape)
print(Y_train.shape)
print(Y_test.shape)

(199364, 29)
(85443, 29)
(199364, 1)
(85443, 1)


# Creating model for Decision Tree Algorithm

In [ ]:
def decision_tree_classification(X_train, Y_train, X_test, Y_test):
    dec_tree_classf = DecisionTreeClassifier()
    print("START MODEL TRAINING ...")
    dec_tree_classf.fit(X_train, Y_train.values.ravel())
    print("COMPLETTION OF MODEL TRAINING")
    accuracy = dec_tree_classf.score(X_test, Y_test)
    print(f'Accuracy : {accuracy}')
    Y_predict = dec_tree_classf.predict(X_test)
    
    #CONFUSION MATRIX
    print(f"Confusion Matrix : \n {confusion_matrix(Y_test, Y_predict)}")
    
    # Classification report for F1_score
    print(f"Classification Report :- \n {classification_report(Y_test, Y_predict)}")

# DECISION TREE CLASSIFICATION MODEL CALLED
decision_tree_classification(X_train, Y_train, X_test, Y_test)

START MODEL TRAINING ...
COMPLETTION OF MODEL TRAINING
Accuracy : 0.9992509626300574
Confusion Matrix : 
 [[85266    30]
 [   34   113]]
Classification Report :- 
               precision    recall  f1-score   support

           0       1.00      1.00      1.00     85296
           1       0.79      0.77      0.78       147

    accuracy                           1.00     85443
   macro avg       0.89      0.88      0.89     85443
weighted avg       1.00      1.00      1.00     85443



# Summary of Results
the scores are converted into percentages with approximate values ---

Accuracy = 99.92%<br>
Precision = 78%<br>
Recall = 76%<br>
F1_Score = 77%<br>